jina/executors/indexers/vector.py

```diff
@@ -137,15 +137,26 @@ def _check_keys(self, keys: Sequence[int]) -> None:
             raise KeyError(f'Key(s) {missed} were not found in {self.save_abspath}')
 
     def update(self, keys: Sequence[int], values: Sequence[bytes], *args, **kwargs) -> None:
-        self.delete(keys)
-        self.add(np.array(keys), np.array(values))
+        keys = filter_keys_nonexistent(set(keys), set(self.ext2int_id.keys()), self.save_abspath, self.logger)
+        # could be empty
+        if keys:
+            # expects np array for computing shapes
+            keys = np.array(list(keys))
+            self.delete(keys, **{'keys_precomputed': True})
+            self.add(np.array(keys), np.array(values))
 
     def delete(self, keys: Sequence[int], *args, **kwargs) -> None:
-        self._check_keys(keys)
-        for key in keys:
-            # mark as `False` in mask
-            self.valid_indices[self.ext2int_id[key]] = False
-            self._size -= 1
+        if kwargs.get('keys_precomputed') is not True:
```

Comment:
the idea would be to have in this class a 

```
def _delete(self, keys, keys_precomputed: bool = False)
```
that is called from different parts

Reply:
We don't even need the `precomputed_keys` arg if we opt for a private `_delete`. The public methods would simply handle the keys, since they are always called first. THis is the way I've implemented it now

Reply:
Also done for the numpyindexer now